source/Octopus.Tentacle/Scripts/Kubernetes/KubernetesJobOutputStreamWriter.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            this.workspace = workspace;
+        }
+
+        public async Task StreamJobLogsToScriptLog(IScriptLogWriter writer, CancellationToken cancellationToken, bool isFinalRead = false)
+        {
+            try
+            {
+                //open the file streams for reading
+                using var stdOutStream = await SafelyOpenLogStreamReader("stdout.log", cancellationToken);
+                using var stdErrStream = await SafelyOpenLogStreamReader("stderr.log", cancellationToken);
+
+                //if either of these is null, just return
+                if (stdOutStream is null || stdErrStream is null)
+                    return;
+
+                while (true)
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                        break;
+
+                    //Read both the stdout and stderr log files
+                    var stdOutReadTask = ReadLogFileTail(writer, stdOutStream, ProcessOutputSource.StdOut, lastStdOutOffset);
+                    var stdErrReadTask = ReadLogFileTail(writer, stdErrStream, ProcessOutputSource.StdErr, lastStdErrOffset);
+
+                    //wait for them to both complete
+                    await Task.WhenAll(stdOutReadTask, stdErrReadTask);
+
+                    //store the final offsets
+                    lastStdOutOffset = stdOutReadTask.Result.FinalOffset;
+                    lastStdErrOffset = stdErrReadTask.Result.FinalOffset;
+
+                    //stitch the log lines together and order by occurred, then write to actual log
+                    var orderedLogLines = stdOutReadTask.Result.Logs
+                        .Concat(stdErrReadTask.Result.Logs)
+                        .OrderBy(ll => ll.Occurred);
+
+                    //write all the read log lines to the output script log
+                    foreach (var logLine in orderedLogLines)
+                    {
+                        writer.WriteOutput(logLine.Source, logLine.Message, logLine.Occurred);
+                    }
+
+                    //wait for 250ms before reading the logs again (except on the final read)
+                    if (!isFinalRead)
+                    {
+                        await Task.Delay(TimeSpan.FromMilliseconds(250), cancellationToken);
+                    }
+                    else
+                    {
+                        //if this is the last read we need to jump out (and not spin forever)
+                        break;
+                    }
+                }
+            }
+            catch (TaskCanceledException)
+            {
+                //ignore all task cancelled exceptions as they may be thrown by the job finishing (and thus signally)
+            }
+        }
+
+        async Task<StreamReader?> SafelyOpenLogStreamReader(string filename, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                if (cancellationToken.IsCancellationRequested)
```

Comment:
I was wondering about the maximum timeout here till I realised we were going to rely on the cancellationToken timeout which makes total sense. Maybe just throw a comment above the `while(true)` about how it can exit since that can raise eyebrows.